src/WebJobs.Script/Workers/Rpc/RpcWorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                // don't allow background exceptions to escape
+                _logger.LogError(ex.ToString());
+            }
+            _timer.Start();
+        }
+
+        internal bool AddWorkerIfNeeded(IEnumerable<IRpcWorkerChannel> workerChannels, TimeSpan elaspsedFromLastAdding)
+        {
+            if (elaspsedFromLastAdding < _concurrencyOptions.Value.AdjustmentPeriod)
+            {
+                return false;
+            }
+
+            bool result = false;
+            IEnumerable<IRpcWorkerChannel> initializedWorkers = workerChannels.Where(ch => ch.IsChannelReadyForInvocations());
+
+            // Check if there are initializing language workers
+            int notInitializedWorkersCount = workerChannels.Count() - initializedWorkers.Count();
+
+            if (notInitializedWorkersCount == 0)
+            {
+                // Check how many channels are oveloaded
+                List<WorkerDescription> descriptions = new List<WorkerDescription>();
+                foreach (IRpcWorkerChannel channel in initializedWorkers)
+                {
+                    if (channel is IRpcWorkerConcurrencyChannel concurrencyChannel)
+                    {
+                        WorkerStatus workerStatus = concurrencyChannel.GetWorkerStatus();
+                        bool overloaded = IsOverloaded(workerStatus);
+                        descriptions.Add(new WorkerDescription()
+                        {
+                            Channel = channel,
+                            WorkerStatus = workerStatus,
+                            Overloaded = overloaded
+                        });
+                    }
+                }
+
+                int overloadedCount = descriptions.Where(x => x.Overloaded == true).Count();
+                if (overloadedCount > 0)
+                {
+                    if (initializedWorkers.Count() < _concurrencyOptions.Value.MaxWorkerCount)
+                    {
+                        _logger.LogDebug($"Adding a new worker, overloaded workers = {overloadedCount}, initialized workers = {initializedWorkers.Count()} ");
+                        result = true;
+                    }
+                }
+
+                if (result == true || _logStateStopWatch.Elapsed > _logStateInterval)
+                {
+                    StringBuilder sb = new StringBuilder();
+                    foreach (WorkerDescription description in descriptions)
+                    {
+                        sb.Append(LogWorkerState(description));
+                        sb.Append(Environment.NewLine);
+                    }
+                    _logStateStopWatch.Restart();
+                    _logger.LogDebug(sb.ToString());
+                }
+            }
+
+            return result;
+        }
+
+        internal bool IsOverloaded(WorkerStatus status)
+        {
+            if (status.RpcWorkerStats.LatencyHistory.Count() >= _concurrencyOptions.Value.HistorySize)
```

Comment:
So only Latency is currently being used to determine whether to scale out workers. I think the right way to do this integration with the new Concurrency Manager stuff would be to:

- implement the WorkerChannelThrottleProvider shell I provide in my host DC PR [here](https://github.com/Azure/azure-functions-host/pull/7406/files#diff-185f0f62183c68e8362f1bdcfe08241606f2917c7874cc740c27c6cda4f0023cR15)
- this throttle provider will query all the channel latencies and determine whether the throttle is Enabled. When it returns it's ConcurrencyThrottleStatus it'll add it's "WorkerChannel" name to the EnabledThrottles collection
- This RpcWorkerConcurrencyService can then take an instance of IConcurrencyThrottleManager on construction, allowing it to check whether the "WorkerChannel" throttle is enabled at any time.

The benefit of putting the latency/throttle logic in the throttle provider is that it's already plugged into ConcurrencyManager, so in addition to you being able to use it for worker scale out, when Enabled it'll also limit function concurrency, so while the channels are overloaded we'll ramp down work. So CM and your new worker scale out logic will work together.

Reply:
The integration with `WorkerChannelThrottleProvider` is planned in the next PR for language worker concurrency.